File: src/schema/flags.ts
import { z } from "zod";

export interface PostHogFeatureFlag {
	id: number;
	key: string;
	name: string;
}

export interface PostHogFlagsResponse {
	results?: PostHogFeatureFlag[];
}
const base = ["exact", "is_not"] as const;
const stringOps = [
	...base,
	"icontains",
	"not_icontains",
	"regex",
	"not_regex",
	"is_cleaned_path_exact",
] as const;
const numberOps = [...base, "gt", "gte", "lt", "lte", "min", "max"] as const;
const booleanOps = [...base] as const;

const arrayOps = ["in", "not_in"] as const;

const operatorSchema = z.enum([
	...stringOps,
	...numberOps,
	...booleanOps,
	...arrayOps,
] as unknown as [string, ...string[]]);

export const PersonPropertyFilterSchema = z
	.object({
		key: z.string(),
		value: z.union([
			z.string(),
			z.number(),
			z.boolean(),
			z.array(z.string()),
			z.array(z.number()),
		]),
		operator: operatorSchema.optional(),
	})
	.superRefine((data, ctx) => {
		const { value, operator } = data;
		if (!operator) return;
		const isArray = Array.isArray(value);

		const valid =
			(typeof value === "string" && stringOps.includes(operator as any)) ||
			(typeof value === "number" && numberOps.includes(operator as any)) ||
			(typeof value === "boolean" && booleanOps.includes(operator as any)) ||
			(isArray && arrayOps.includes(operator as any));

		if (!valid)
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				message: `operator "${operator}" is not valid for value type "${isArray ? "array" : typeof value}"`,
			});

		if (!isArray && arrayOps.includes(operator as any))
			ctx.addIssue({
				code: z.ZodIssueCode.custom,
				message: `operator "${operator}" requires an array value`,
			});
	})
	.transform((data) => {
		return {
			...data,
			type: "person",
		};
	});

export type PersonPropertyFilter = z.infer<typeof PersonPropertyFilterSchema>;

export const FiltersSchema = z.object({
	properties: z.array(PersonPropertyFilterSchema),
	rollout_percentage: z.number(),
});

export type Filters = z.infer<typeof FiltersSchema>;

export const FilterGroupsSchema = z.object({
	groups: z.array(FiltersSchema),
});

export type FilterGroups = z.infer<typeof FilterGroupsSchema>;

export const CreateFeatureFlagInputSchema = z.object({
	name: z.string(),
	key: z.string(),
	description: z.string(),
	filters: FilterGroupsSchema,
	active: z.boolean(),
	tags: z.array(z.string()).optional(),
});

export type CreateFeatureFlagInput = z.infer<typeof CreateFeatureFlagInputSchema>;

export const UpdateFeatureFlagInputSchema = CreateFeatureFlagInputSchema.omit({
	key: true,
}).partial();

export type UpdateFeatureFlagInput = z.infer<typeof UpdateFeatureFlagInputSchema>;

export const FeatureFlagSchema = z.object({
	id: z.number(),
	key: z.string(),
	name: z.string(),
	description: z.string().optional(),
	filters: FiltersSchema.optional(),
	active: z.boolean(),
	tags: z.array(z.string()).optional(),
});

export type FeatureFlag = z.infer<typeof FeatureFlagSchema>;

File: python/schema/flags.py
from enum import Enum

from pydantic import BaseModel, model_validator


class PostHogFeatureFlag(BaseModel):
    id: int
    key: str
    name: str


class PostHogFlagsResponse(BaseModel):
    results: list[PostHogFeatureFlag] | None = None


class OperatorType(str, Enum):
    EXACT = "exact"
    IS_NOT = "is_not"
    ICONTAINS = "icontains"
    NOT_ICONTAINS = "not_icontains"
    REGEX = "regex"
    NOT_REGEX = "not_regex"
    IS_CLEANED_PATH_EXACT = "is_cleaned_path_exact"
    GT = "gt"
    GTE = "gte"
    LT = "lt"
    LTE = "lte"
    MIN = "min"
    MAX = "max"
    IN = "in"
    NOT_IN = "not_in"


class PersonPropertyFilter(BaseModel):
    key: str
    value: str | int | bool | list[str] | list[int]
    operator: OperatorType | None = None
    type: str = "person"

    @model_validator(mode="after")
    def validate_operator_value_compatibility(self):
        if not self.operator:
            return self

        value = self.value
        operator = self.operator
        is_array = isinstance(value, list)

        string_ops = {
            OperatorType.EXACT,
            OperatorType.IS_NOT,
            OperatorType.ICONTAINS,
            OperatorType.NOT_ICONTAINS,
            OperatorType.REGEX,
            OperatorType.NOT_REGEX,
            OperatorType.IS_CLEANED_PATH_EXACT,
        }
        number_ops = {
            OperatorType.EXACT,
            OperatorType.IS_NOT,
            OperatorType.GT,
            OperatorType.GTE,
            OperatorType.LT,
            OperatorType.LTE,
            OperatorType.MIN,
            OperatorType.MAX,
        }
        boolean_ops = {OperatorType.EXACT, OperatorType.IS_NOT}
        array_ops = {OperatorType.IN, OperatorType.NOT_IN}

        valid = (
            (isinstance(value, str) and operator in string_ops)
            or (isinstance(value, int | float) and operator in number_ops)
            or (isinstance(value, bool) and operator in boolean_ops)
            or (is_array and operator in array_ops)
        )

        if not valid:
            value_type = "array" if is_array else type(value).__name__
            raise ValueError(f'operator "{operator}" is not valid for value type "{value_type}"')

        if not is_array and operator in array_ops:
            raise ValueError(f'operator "{operator}" requires an array value')

        return self


class Filters(BaseModel):
    properties: list[PersonPropertyFilter]
    rollout_percentage: int


class FilterGroups(BaseModel):
    groups: list[Filters]


class CreateFeatureFlagInput(BaseModel):
    name: str
    key: str
    description: str
    filters: FilterGroups
    active: bool
    tags: list[str] | None = None


class UpdateFeatureFlagInput(BaseModel):
    name: str | None = None
    description: str | None = None
    filters: FilterGroups | None = None
    active: bool | None = None
    tags: list[str] | None = None


class FeatureFlag(BaseModel):
    id: int
    key: str
    name: str
    description: str | None = None
    filters: Filters | None = None
    active: bool
    tags: list[str] | None = None
